test/framework/deployment_helpers.go

```diff
@@ -179,51 +183,111 @@ type watchPodLogsInput struct {
-				Namespace: input.Namespace,
-				App:       input.DeploymentName,
-				Pod:       pod.Name,
-				Container: container.Name,
-				NodeName:  pod.Spec.NodeName,
-				Stream:    "stderr",
-			}
-			metadataBytes, err := json.Marshal(&metadata)
-			Expect(err).To(BeNil())
-			Expect(os.WriteFile(logMetadataFile, metadataBytes, 0600)).To(Succeed())
+		// Create log metadata file.
+		logMetadataFile := filepath.Clean(path.Join(input.LogPath, input.DeploymentName, pod.Name, container.Name+"-log-metadata.json"))
+		Expect(os.MkdirAll(filepath.Dir(logMetadataFile), 0750)).To(Succeed())
 
-			// Watch each container's logs in a goroutine so we can stream them all concurrently.
-			go func(pod corev1.Pod, container corev1.Container) {
-				defer GinkgoRecover()
+		metadata := logMetadata{
+			Job:       input.Namespace + "/" + input.DeploymentName,
+			Namespace: input.Namespace,
+			App:       input.DeploymentName,
+			Pod:       pod.Name,
+			Container: container.Name,
+			NodeName:  pod.Spec.NodeName,
+			Stream:    "stderr",
+		}
+		metadataBytes, err := json.Marshal(&metadata)
+		Expect(err).To(BeNil())
+		Expect(os.WriteFile(logMetadataFile, metadataBytes, 0600)).To(Succeed())
 
-				logFile := filepath.Clean(path.Join(input.LogPath, input.DeploymentName, pod.Name, container.Name+".log"))
-				Expect(os.MkdirAll(filepath.Dir(logFile), 0750)).To(Succeed())
+		// Watch each container's logs in a goroutine so we can stream them all concurrently.
+		go func(pod corev1.Pod, container corev1.Container) {
+			defer GinkgoRecover()
 
-				f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
-				Expect(err).NotTo(HaveOccurred())
-				defer f.Close()
+			logFile := filepath.Clean(path.Join(input.LogPath, input.DeploymentName, pod.Name, container.Name+".log"))
+			Expect(os.MkdirAll(filepath.Dir(logFile), 0750)).To(Succeed())
 
-				opts := &corev1.PodLogOptions{
-					Container: container.Name,
-					Follow:    true,
-				}
+			f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
+			Expect(err).NotTo(HaveOccurred())
+			defer f.Close()
+
+			opts := &corev1.PodLogOptions{
+				Container: container.Name,
+				Follow:    true,
+			}
 
-				podLogs, err := input.ClientSet.CoreV1().Pods(input.Namespace).GetLogs(pod.Name, opts).Stream(ctx)
+			// Retry streaming the logs of the pods unless ctx.Done() or if the pod does not exist anymore.
+			err = wait.PollInfiniteWithContext(ctx, 2*time.Second, func(ctx context.Context) (done bool, err error) {
+				// Wait for pod to be in running state
+				actual, err := input.ClientSet.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
 				if err != nil {
+					if apierrors.IsNotFound(err) {
+						return true, nil
+					}
+					// Failing to get the pod should not cause the test to fail
+					log.Logf("Error getting pod %s, container %s: %v", klog.KRef(pod.Namespace, pod.Name), container.Name, err)
+					return true, nil
```

Comment:
Wondering if we want to retry in this case

Reply:
My thought here (as far as I can remember): its in a `wait.PollInfinite`, so if the error would not go away, it would retry forever 🤔 

however: if we don't want to retry, we should return the error instead, because it would also only log the error.

Reply:
Makes sense. Let's go with your current implementation. We get a log and we can improve over time if we see that we want to retry other **specific** errors.